sw/device/lib/testing/spi_flash_testutils.c

```diff
@@ -205,3 +248,142 @@ void spi_flash_testutils_program_page(dif_spi_host_t *spih, uint8_t *payload,
 }
+
+void spi_flash_testutils_program_page(dif_spi_host_t *spih, uint8_t *payload,
+                                      size_t length, uint32_t address,
+                                      bool addr_is_4b) {
+  spi_flash_testutils_program_op(spih, kSpiDeviceFlashOpPageProgram, payload,
+                                 length, address, addr_is_4b);
+}
+
+void spi_flash_testutils_read_op(dif_spi_host_t *spih, uint8_t opcode,
+                                 uint8_t *payload, size_t length,
+                                 uint32_t address, bool addr_is_4b,
+                                 uint8_t width, uint8_t dummy) {
+  CHECK(spih != NULL);
+  CHECK(payload != NULL);
+  CHECK(length <= 256);  // Length must be less than a page size.
+  CHECK(width == 1 || width == 2 || width == 4);
+
+  dif_spi_host_addr_mode_t addr_mode =
+      addr_is_4b ? kDifSpiHostAddrMode4b : kDifSpiHostAddrMode3b;
+  dif_spi_host_width_t data_width = width == 1   ? kDifSpiHostWidthStandard
+                                    : width == 2 ? kDifSpiHostWidthDual
+                                                 : kDifSpiHostWidthQuad;
+
+  dif_spi_host_segment_t transaction[4] = {
+      {
+          .type = kDifSpiHostSegmentTypeOpcode,
+          .opcode = opcode,
+      },
+      {
+          .type = kDifSpiHostSegmentTypeAddress,
+          .address =
+              {
+                  .width = kDifSpiHostWidthStandard,
+                  .mode = addr_mode,
+                  .address = address,
+              },
+      },
+      {
+          .type = kDifSpiHostSegmentTypeDummy,
+          .dummy =
+              {
+                  .width = kDifSpiHostWidthStandard,
+                  .length = dummy,
+              },
+      },
+      {
+          .type = kDifSpiHostSegmentTypeRx,
+          .rx =
+              {
+                  .width = data_width,
+                  .buf = payload,
+                  .length = length,
+              },
+      },
+  };
+  size_t tlen = ARRAYSIZE(transaction);
+  if (dummy == 0) {
+    // If there are no dummy cycles, then get rid of the dummy segment in
+    // the transaction.
+    memcpy(&transaction[2], &transaction[3], sizeof(dif_spi_host_segment_t));
+    tlen -= 1;
+  }
+  CHECK_DIF_OK(dif_spi_host_transaction(spih, /*csid=*/0, transaction, tlen));
+}
+
+status_t spi_flash_testutils_quad_enable(dif_spi_host_t *spih, uint8_t method,
+                                         bool enabled) {
+  uint32_t status;
+  switch (method) {
+    case 0:
+      // Device does not have a QE bit.
+      break;
+    case 1:
+      // QE is bit1 of status reg 2.
+      // Set/clear via two-byte reads/writes via SR1 opcodes.
+      // Writing only one byte to SR1 clears SR2.
+      status = TRY(spi_flash_testutils_read_status(
+          spih, kSpiDeviceFlashOpReadStatus1, 2));
+      status = bitfield_bit32_write(status, 9, enabled);
+      status = TRY(spi_flash_testutils_write_status(
```

Comment:
`status` is not used later, so maybe:
```suggestion
    TRY(spi_flash_testutils_write_status(
```